core/src/main/java/amino/run/oms/migrationdecision/MetricWatcher.java

```diff
@@ -0,0 +1,282 @@
+                e.printStackTrace();
+            } catch (MicroServiceReplicaNotFoundException e) {
+                logger.warning(
+                        String.format(
+                                "Skipping Replica [%s] evaluation as micro service replica not available",
+                                replicaID));
+                e.printStackTrace();
+            } catch (KernelServerNotFoundException e) {
+                logger.warning(
+                        String.format(
+                                "Skipping Replica [%s] evaluation as micro service replica's kernel server not available",
+                                replicaID));
+                e.printStackTrace();
+            }
+        }
+
+        // add and print best replica and their processing time in sorted order
+        for (AbstractMap.SimpleEntry<ReplicaID, Long> replicaStat : replicaHeap) {
+            logger.fine(String.format("Replica status : [%s]", replicaStat));
+            if (bestCandidates.size() <= MAX_TOP_CANDIDATE) {
+                bestCandidates.add(replicaStat.getKey());
+            }
+        }
+
+        // clear replica heap
+        replicaHeap.clear();
+        return bestCandidates;
+    }
+
+    private boolean triggerMigration(ReplicaID replicaID, InetSocketAddress kernelServer) {
+        MicroServiceID microServiceID = replicaID.getOID();
+
+        // start migration
+        try {
+            KernelOID koid = microServiceManager.getRootGroupId(microServiceID);
+            EventHandler groupPolicyEventHandler =
+                    microServiceManager.getGroupDispatcher(microServiceID, koid);
+            DefaultPolicy.DefaultGroupPolicy replicaGroupPolicy =
+                    (DefaultPolicy.DefaultGroupPolicy) groupPolicyEventHandler.getObjects().get(0);
+            replicaGroupPolicy.onNotification(new MigrationNotification(replicaID, kernelServer));
+        } catch (MicroServiceNotFoundException e) {
+            logger.warning(
+                    String.format(
+                            "Failed to migrate replica [%s] as micro service not available",
+                            replicaID));
+            e.printStackTrace();
+            return false;
+        } catch (RemoteException e) {
+            logger.warning(
+                    String.format(
+                            "Failed to migrate replica [%s] with remote exception : %s",
+                            replicaID, e.getMessage()));
+            e.printStackTrace();
+            return false;
+        }
+
+        return true;
+    }
+
+    private long aggregatedProcessingTime(ReplicaID replicaID)
+            throws MicroServiceNotFoundException, MicroServiceReplicaNotFoundException,
+                    KernelServerNotFoundException {
+        /*
+           To measure time consumption by vertex Vk we use the following formula
+           tk = Tk,posk+i + summation { (Ski+Sik)/Rposk,posi + Lposk,posi }
```

Comment:
Please explain what all these terms mean.  tK, Tk, posk+i, Ski, Sik, Rposk?  It's not at all clear what you mean here.